core/domain/html_validation_service.py

```diff
@@ -820,54 +819,163 @@ def regenerate_image_filename_using_dimensions(filename, height, width):
+        tab_component['tab_contents-with-value'] = (
+            escape_html(json.dumps(tab_content_dict_list)).replace(
+                '\\n', '\n'))
+    return python_utils.UNICODE(soup).replace('<br/>', '<br>')
+
+
+def _add_dimensions_to_image_filenames(exploration_fs, soup):
+    """Modifies filenames of images. This is a helper method for
+    add_dims_to_img_in_complex_rte.
+
+    Args:
+        exploration_fs: GcsFileSystem. File system of the Exploration entity.
+        soup: bs4.BeautifulSoup. The html soup whose image file is
+            to be renamed.
+    """
+    for image in soup.findAll(name='oppia-noninteractive-image'):
+        if (not image.has_attr('filepath-with-value') or
+                image['filepath-with-value'] == '' or
+                image['filepath-with-value'] == '""' or
+                image['filepath-with-value'] == '&quot;&quot;'):
             image.decompose()
             continue
 
         try:
-            filename = json.loads(unescape_html(image['filepath-with-value']))
-            image['filepath-with-value'] = escape_html(json.dumps(
-                get_filename_with_dimensions(filename, exp_id)))
+            filename = (
+                json.loads(unescape_html(image['filepath-with-value'])))
+            escaped_filename = escape_html(
+                json.dumps(
+                    get_filename_with_dimensions(exploration_fs, filename)))
+            image['filepath-with-value'] = escaped_filename
         except Exception as e:
+            (entity_type, entity_id, _) = (
+                exploration_fs.impl.assets_path.split('/'))
             logging.error(
-                'Exploration %s failed to load image: %s' %
-                (exp_id, image['filepath-with-value'].encode('utf-8')))
+                '%s %s failed to load image: %s' % (
+                    entity_type, entity_id,
+                    image['filepath-with-value'].encode('utf-8')))
             raise e
-    return python_utils.UNICODE(soup).replace('<br/>', '<br>')
 
 
-def get_filename_with_dimensions(old_filename, exp_id):
+def get_filename_with_dimensions(exploration_fs, old_filename):
     """Gets the filename with dimensions of the image file in it.
 
     Args:
+        exploration_fs: GcsFileSystem. File system of the Exploration entity.
         old_filename: str. Name of the file whose dimensions need to be
             calculated.
-        exp_id: str. Exploration id.
 
     Returns:
         str. The new filename of the image file.
     """
-    file_system_class = fs_services.get_entity_file_system_class()
-    fs = fs_domain.AbstractFileSystem(file_system_class(
-        feconf.ENTITY_TYPE_EXPLORATION, exp_id))
-    filepath = 'image/%s' % old_filename
+    subdirectory = 'image'
+    filepath = '%s/%s' % (subdirectory, old_filename)
     try:
-        content = fs.get(filepath.encode('utf-8'))
+        dot_index = old_filename.rfind('.')
```

Comment:
Hi @seanlip, I've made some changes here. PTAL!

Reply:
I think you may want to include some comments/validation about what the format of the old_filename is? You're making some assumptions about its structure when computing the filename_regex but these assumptions should be validated before that.

Reply:
Would something along these lines be better?

```python
        # If the old_filename does not contain the pattern
        # _height_<number>_width_<number>, it is assumed that the
        # old_filename was not updated. In this case, the following code
        # iterates over the image subdirectory in the datastore and
        # looks for a filename with the pattern
        # <old_filename>_height_<number>_width_<number>.<extension>.
        # If such a filename is found, it is returned.
        if re.match(r'^((?!_height_\d+_width_\d+).)*$', old_filename):
            dot_index = old_filename.rfind('.')
            filename_regex = r'%s/%s_height_[0-9]+_width_[0-9]+\.\w+' % (
                subdirectory, old_filename[:dot_index])
            for filename in exploration_fs.listdir(
                    subdirectory.encode(encoding='utf-8')):
                if re.match(filename_regex, filename):
                    (_, new_filename) = filename.split('/')
                    return new_filename
        else:
            raise Exception(
                'old_filename must not contain the pattern _height_<number>'
                '_width_<number>')
```

Reply:
I think something along these lines is reasonable, yep -- as long as you ensure that callers don't call this function with such filenames and update the docstring to say what filenames are allowed!

Reply:
Done. Thanks!